jaxrs/src/main/java/io/airlift/jaxrs/JacksonJsonMapper.java

```diff
@@ -42,15 +46,18 @@ public JsonMapper(ObjectMapper objectMapper)
      */
     @Override
     public Object readFrom(Class<Object> type, Type genericType, Annotation[] annotations, MediaType mediaType, MultivaluedMap<String, String> httpHeaders, InputStream entityStream)
-            throws IOException
+            throws JacksonException
     {
         try {
             return super.readFrom(type, genericType, annotations, mediaType, httpHeaders, entityStream);
         }
-        catch (IOException e) {
+        catch (JacksonException | NoContentException e) {
             // Re-throw real IO exceptions that are not due to bad JSON
-            if (!(e instanceof JsonProcessingException) && !(e instanceof EOFException)) {
-                throw e;
+            if (e instanceof JacksonIOException ioException) {
+                throw new UncheckedIOException(ioException.getCause());
+            }
+            if (e instanceof StreamReadException e2) {
+                throw e2;
             }
             throw new JsonParsingException(e);
         }
```

Comment:
The confusingly named `StreamReadException` seems to be used for invalid JSON, not IO errors, so we should convert it. I think this is correct:

```suggestion
        catch (JacksonIOException e) {
            // Re-throw real IO exceptions that are not due to bad JSON
            throw e;
        }
        catch (JacksonException | NoContentException e) {
            throw new JsonParsingException(e);
        }
```